### PR TITLE
lxcfs: massage locking

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: src/lxcfs.c [hunk 1]
```diff
@@ -157,18 +159,26 @@ static void do_reload(void)
...
 {
+	int nr_users;
+
+	nr_users = atomic_fetch_inc(&users_count);
+	if (!need_reload)
+		return;
+
 	users_lock();
-	if (users_count == 0 && need_reload)
+	if (nr_users == 0) {
```
Comment (Blub) [inline src/lxcfs.c:169]: I think the `nr_users == 0` check might fail if there were ongoing requests when the first thread after `need_reload` became `true` enters this.

I think we need to decrement after the first `if` again and introduce a second condvar guarding the `do_reload()` which gets triggered in `down_user` if the count drops to 0 and `need_reload` is set?

File: src/lxcfs.c [hunk 2]
```diff
@@ -157,18 +159,26 @@ static void do_reload(void)
...
 		do_reload();
-	users_count++;
+                pthread_cond_signal(&var_cond);
```
Comment (Blub) [inline src/lxcfs.c:171]: I think this should be `pthread_cond_broadcast()` to wake up *all* waiters.

File: src/lxcfs.c (hunk 1 above)
Comment (brauner) [reply src/lxcfs.c:169]: Hm, `nr_users` is retrieved via `atomic_fetch_inc()` so it shouldn't change anymore after we've retrieved it.

File: src/lxcfs.c [hunk 3]
```diff
@@ -149,31 +151,41 @@ static void do_reload(void)
...
 }
 
 static void up_users(void)
 {
-	users_lock();
-	if (users_count == 0 && need_reload)
-		do_reload();
-	users_count++;
-	users_unlock();
+	if (atomic_load_acquire(&need_reload)) {
```
Comment (Blub) [inline src/lxcfs.c:158]: Doing this before the `increment` can be dangerous, though.
Assume thread 1 goes to the *top* of the `else` branch before incrementing, `user_count` is still `0`.
Then `need_reload` gets set and another thread goes into this branch right up to the `do_reload()`, and right in the middle of it, the scheduler switches back to thread 1 which bumps the counter and continues towards a crash.

Comment (brauner) [reply src/lxcfs.c:158]: Yeah, I already had deadlocks anyway.

**Pull request closed without merging**